fixes signature verification in verifylib

RohanBhirangi opened the pull request:
verify all steps signatures strictly checks if the link is signed
by the keyid mentioned in the filename

RohanBhirangi committed:
fixes signature verification in verifylib

verify all steps signatures strictly checks if the link is signed
by the keyid mentioned in the filename

SantiagoTorres commented:
Hi @RohanBhirangi , Maybe I"m missing something but I don't see this is using the filename anywhere?

RohanBhirangi commented:
Hey, we load the link files (step.keyid.link) into the chain_link_dict ({step:{keyid: link}}) and check whether the link has been signed by the key corresponding to the keyid.
I assume this is what we wanted.

lukpueh commented on in_toto/verifylib.py:
```diff
@@ -239,16 +239,17 @@ def verify_all_steps_signatures(layout, chain_link_dict):
+      # Create the dictionary of keys for this step
+      for pubkeyid in step.pubkeys:
+          if pubkeyid == keyid:
+            keys_dict[pubkeyid] = layout.keys[pubkeyid]
```
Here are some notes,
- make sure to always have consistent indentation (2 spaces per level, 4 for line continuation),
- use `if pubkeyid in step.pubkeys` instead of the loop,
- I think it makes sense to fail early (plus give user feedback) , e.g. something like:
```python
if pubkeyid in step.pubkeys:
  keys_dict[pubkeyid] = layout.keys[pubkeyid]
else:
  raise SignatureVerificationError("<some descriptive message here>")
```
- also remove the signature verification part from the [threshold verification function]( https://github.com/in-toto/in-toto/blob/develop/in_toto/verifylib.py#L986-L993) as mentioned in the related [issue](https://github.com/in-toto/in-toto/issues/82),
- try to add some comments that describe what your changes are good for,
- if you fix an issue please mention the issue in the PR.

SantiagoTorres replied:
I'm adding to this, where does the keyid prefix of the link filename check actually happen?  I'm finding it hard to parse.

lukpueh replied:
The variable value `keyid` holds the (whole) keyid, determined by the filename infix. So instead of taking all keys authorized for a certain step (`step.pubkeys`) and verifying that the link file is signed by one ore more of those keys, we take the key, identified by `keyid`, check if it is authorized, and then verify if the link was signed with that key.

RohanBhirangi replied:
> if pubkeyid in step.pubkeys:
  keys_dict[pubkeyid] = layout.keys[pubkeyid]
else:
  raise SignatureVerificationError("<some descriptive message here>")


@lukpueh  Do we need this? This will raise an error if the keyid (as in the filename) is not present in the layout. But if it isn't present in the layout, our initial load_metadata will fail.
Or maybe I'm missing something.

lukpueh replied:
Isn't this check something you are [proposing in your PR](https://github.com/RohanBhirangi/in-toto/blob/225c0169f672fa16972ba893c67a39b149313343/in_toto/verifylib.py#L249-L250)? 

The only difference is that you would let a non-authorized key fail implicitly in the `Signable`'s `verify_signature` function due to an [empty signatures field](https://github.com/in-toto/in-toto/blob/develop/in_toto/models/common.py#L102) or an [empty key dictionary](https://github.com/in-toto/in-toto/blob/develop/in_toto/models/common.py#L108), and I suggest to let it fail explicitly. 

Maybe `SignatureVerificationError` is not the best exception type, maybe it should be something like `AuthorizationError`. 

In any case I'd say it does not hurt to re-check the authorization here, although it's already done in the `import_step_metadata_from_files_as_dict` function.

lukpueh reviewed (commented):
Thanks for the PR, please address comments.

RohanBhirangi committed:
modified signature verification of link files

added AuthorizationError to in_toto.exceptions
removed loop for finding the keyid
removed duplicate signature verification from verify_threshold()

lukpueh commented on in_toto/exceptions.py:
```diff
@@ -25,3 +25,8 @@ class LinkNotFoundError(Error):
+
+class AuthorizationError(Error):
+  """Indicates that a ran command exited with non-int or non-zero return
+  value. """
```
Please update the docstrings when you copy-paste. ;)

lukpueh commented on in_toto/verifylib.py:
```diff
@@ -246,9 +246,13 @@ def verify_all_steps_signatures(layout, chain_link_dict):
+      # filename is added to the dictionary which ensures
+      # that the link has been signed by that key
+      if keyid in step.pubkeys:
+            keys_dict[keyid] = layout.keys[keyid]
```
Indentation!

lukpueh reviewed (changes requested):
Please fix tiny issues and we can merge.

RohanBhirangi committed:
changed AuthorizationError docstring in in_toto/exceptions.py

fixed indentation in verifylib.py
changed the docstring of LinkNotFoundError in exceptions.py

SantiagoTorres commented:
this LGTM
